### PR TITLE
[spirv] support clip/cull distance stage vars with array types

### LINKED ISSUE #3278: [SPIR-V] SV_ClipDistance variable not allowed to be of array type (works with DXIL backend and FXC)
### Title
SV_ClipDistance variable not allowed to be of array type (works with DXIL backend and FXC)

### Functional impact
Breaks compatibility with DXIL backend and FXC. The issue appears to have been fixed for the DXIL backend some time ago https://github.com/microsoft/DirectXShaderCompiler/issues/2057 but not for the SPIR-V backend.

### Minimal repro steps

1. Compile the following shader with `dxc.exe -T vs_6_0 -E VSMain issue_shader.txt -spirv`:
```
struct v2f {
	float4	vertex		: SV_Position;
	float3	clipDistance[2]: SV_ClipDistance;
};

v2f VSMain (float4 vertex : POSITION) {
	v2f o;
        o.vertex = vertex;

	float c = vertex.y;
	o.clipDistance[0].xyz = float3(c, c+1.0, c+2.0);
	o.clipDistance[1].xyz = float3(c+4.0, c+5.0, c+6.0);
	return o;
}
```
2. Observe compiler output:
```
SV_ClipDistance_repro.txt:3:2: error: SV_ClipDistance variable 'clipDistance' not allowed to be of array type
        float3  clipDistance[2]: SV_ClipDistance;
        ^ 
```
### Expected result
Compilation succeeds like it does for the DXIL backen (i.e with no `-spirv` argument)

### Actual result
Compilation fails with the error above

### Further technical details
Tested with DXC built from source 22dc78f4c3f745c82710fd69f5069e02b5131cae (Nov 16 2020)

### PR DESCRIPTION
**Pull request by jaebaek**
`GlPerVertex` class mainly handles clip/cull distance stage variable.

This change entirely refactors methods of `GlPerVertex` class to
improve the code readability. Mainly it splits each method handling
multiple cases within a single method/loop into multiple methods.

In addition, this change adds the code to support  clip/cull distance
stage vars with array types.

Fixes #3278

### CODE REVIEW

File: tools/clang/lib/SPIRV/GlPerVertex.cpp [hunk 1]
```diff
@@ -126,6 +126,202 @@ bool GlPerVertex::recordGlPerVertexDeclFacts(const DeclaratorDecl *decl,
   return doGlPerVertexFacts(decl, type, asInput);
 }
 
+bool GlPerVertex::isArrayVectorOrScalarOnlyWithFloats(QualType type) const {
```
Comment (sudonatalie) [inline tools/clang/lib/SPIRV/GlPerVertex.cpp:129]: Would is still be correct if we simplified to `isOrContainsFloatType`?

File: tools/clang/lib/SPIRV/GlPerVertex.cpp [hunk 2]
```diff
@@ -233,52 +429,40 @@ bool GlPerVertex::doGlPerVertexFacts(const DeclaratorDecl *decl,
   if (baseType->isReferenceType())
     baseType = baseType->getPointeeType();
 
-  if (baseType->isFloatingType() || hlsl::IsHLSLVecType(baseType)) {
-    (*typeMap)[semanticIndex] = baseType;
-    return true;
+  // Clip/cull distance must be made up only with floats.
+  if (!isArrayVectorOrScalarOnlyWithFloats(baseType)) {
+    emitError("elements for %select{SV_ClipDistance|SV_CullDistance}0 "
+              "variable '%1' must be scalar, vector, or array with floats",
```
Comment (sudonatalie) [inline tools/clang/lib/SPIRV/GlPerVertex.cpp:435]: nit: "of floats" or "of float type"

File: tools/clang/lib/SPIRV/GlPerVertex.cpp [hunk 3]
```diff
@@ -233,52 +429,40 @@ bool GlPerVertex::doGlPerVertexFacts(const DeclaratorDecl *decl,
...
+    if (*blockArraySize ==
+        static_cast<uint32_t>(arrayType->getSize().getZExtValue())) {
+      if (setClipCullDistanceType(typeMap, semanticIndex,
+                                  arrayType->getElementType())) {
         return true;
       }
 
       emitError("elements for %select{SV_ClipDistance|SV_CullDistance}0 "
-                "variable '%1' must be (vector of) floats",
+                "variable '%1' must be scalar, vector, or array with floats",
```
Comment (sudonatalie) [inline tools/clang/lib/SPIRV/GlPerVertex.cpp:453]: same here

File: tools/clang/lib/SPIRV/GlPerVertex.cpp [hunk 4]
```diff
@@ -233,52 +429,40 @@ bool GlPerVertex::doGlPerVertexFacts(const DeclaratorDecl *decl,
...
-        << isCull << decl->getName();
-    return false;
+  if (setClipCullDistanceType(typeMap, semanticIndex, baseType)) {
+    return true;
   }
 
-  emitError("incorrect type for %select{SV_ClipDistance|SV_CullDistance}0 "
-            "variable '%1'",
+  emitError("type for %select{SV_ClipDistance|SV_CullDistance}0 "
+            "variable '%1' must be a scalar, vector, or array with floats",
```
Comment (sudonatalie) [inline tools/clang/lib/SPIRV/GlPerVertex.cpp:465]: and here

File: tools/clang/lib/SPIRV/GlPerVertex.cpp [hunk 5]
```diff
@@ -305,18 +490,13 @@ void GlPerVertex::calculateClipCullDistanceArraySize() {
...
-      } else if (isVectorType(type, &elemType, &count)) {
-        (*offsetMap)[index] = *totalSize;
-        *totalSize += count;
-      } else {
-        llvm_unreachable("SV_ClipDistance/SV_CullDistance not float or "
-                         "vector of float case sneaked in");
+      uint32_t count = getNumberOfScalarComponentsInScalarVectorArray(type);
+      if (count == 0) {
+        llvm_unreachable("SV_ClipDistance/SV_CullDistance not scalar, vector, "
+                         "or array of floats case sneaked in");
```
Comment (sudonatalie) [inline tools/clang/lib/SPIRV/GlPerVertex.cpp:496]: I don't quite understand this message, should it be something like "SV_ClipDistance/SV_CullDistance must be non-empty"

File: tools/clang/lib/SPIRV/GlPerVertex.cpp [hunk 6]
```diff
@@ -406,99 +586,28 @@ bool GlPerVertex::tryToAccess(hlsl::SigPoint::Kind sigPointKind,
 SpirvInstruction *GlPerVertex::readClipCullArrayAsType(
     bool isClip, uint32_t offset, QualType asType, SourceLocation loc) const {
   SpirvVariable *clipCullVar = isClip ? inClipVar : inCullVar;
-
-  // The ClipDistance/CullDistance is always an float array. We are accessing
-  // it using pointers, which should be of pointer to float type.
-  const QualType f32Type = astContext.FloatTy;
+  uint32_t count = getNumberOfScalarComponentsInScalarVectorArray(asType);
+  if (count == 0) {
+    llvm_unreachable("SV_ClipDistance/SV_CullDistance not scalar, vector, "
```
Comment (sudonatalie) [inline tools/clang/lib/SPIRV/GlPerVertex.cpp:591]: Same here

**Review by sudonatalie — APPROVED** [review]
LGTM! Just optional nits in messaging/naming

File: tools/clang/lib/SPIRV/GlPerVertex.cpp (hunk 2 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/GlPerVertex.cpp]: done

File: tools/clang/lib/SPIRV/GlPerVertex.cpp (hunk 3 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/GlPerVertex.cpp]: done

File: tools/clang/lib/SPIRV/GlPerVertex.cpp (hunk 4 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/GlPerVertex.cpp]: done

File: tools/clang/lib/SPIRV/GlPerVertex.cpp (hunk 5 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/GlPerVertex.cpp]: done

File: tools/clang/lib/SPIRV/GlPerVertex.cpp (hunk 6 above)
Comment (jaebaek) [reply tools/clang/lib/SPIRV/GlPerVertex.cpp]: done

**Review by jaebaek — COMMENTED** [review]
Thank you for the feedback!

**Pull request merged**